pallets/delegation/src/benchmarking.rs

```diff
@@ -0,0 +1,373 @@
+	hash.into()
+}
+
+/// sets parent to `None` if it is the root
+fn parent_id_check<T: Config>(
+	root_id: T::DelegationNodeId,
+	parent_id: T::DelegationNodeId,
+) -> Option<T::DelegationNodeId> {
+	if parent_id == root_id {
+		None
+	} else {
+		Some(parent_id)
+	}
+}
+
+/// add ctype to storage and root delegation
+fn add_root_delegation<T: Config>(
+	number: u32,
+) -> Result<(sr25519::Public, T::AccountId, T::DelegationNodeId, T::Hash), DispatchError>
+where
+	T::AccountId: From<sr25519::Public>,
+	T::DelegationNodeId: From<T::Hash>,
+{
+	let root_public = sr25519_generate(KeyTypeId(*b"aura"), None);
+	let root_acc: T::AccountId = root_public.into();
+	let ctype_hash = <T::Hash as Default>::default();
+	let root_id = generate_delegation_id::<T>(number);
+
+	ctype::Module::<T>::add(RawOrigin::Signed(root_acc.clone()).into(), ctype_hash)?;
+	Module::<T>::create_root(
+		RawOrigin::Signed(root_acc.clone()).into(),
+		root_id,
+		ctype_hash,
+	)?;
+
+	Ok((root_public, root_acc, root_id, ctype_hash))
+}
+
+/// recursively adds children delegations to a parent delegation for each level until reaching leaf level
+fn add_children<T: Config>(
+	root_id: T::DelegationNodeId,
+	parent_id: T::DelegationNodeId,
+	parent_acc_public: sr25519::Public,
+	parent_acc_id: T::AccountId,
+	permissions: Permissions,
+	level: u32,
+	children_per_level: u32,
+) -> Result<(sr25519::Public, T::AccountId, T::DelegationNodeId), DispatchError>
+where
+	T::AccountId: From<sr25519::Public>,
+	T::Signature: From<sr25519::Signature>,
+	T::DelegationNodeId: From<T::Hash>,
+{
+	if level == 0 {
+		return Ok((parent_acc_public, parent_acc_id, parent_id));
+	};
+
+	let mut leaf = None;
+	for c in 0..children_per_level {
+		// setup delegation account and id
+		let delegation_acc_public = sr25519_generate(KeyTypeId(*b"aura"), None);
+		let delegation_acc_id: T::AccountId = delegation_acc_public.into();
+		let delegation_id = generate_delegation_id::<T>(level * children_per_level + c);
+
+		// only set parent if not root
+		let parent = parent_id_check::<T>(root_id, parent_id);
+
+		// delegate signs delegation to parent
+		let hash: Vec<u8> =
+			Module::<T>::calculate_hash(delegation_id, root_id, parent, permissions).encode();
+		let sig: T::Signature =
+			sp_io::crypto::sr25519_sign(KeyTypeId(*b"aura"), &delegation_acc_public, hash.as_ref())
+				.ok_or("Error while building signature of delegation.")?
+				.into();
+
+		// add delegation from delegate to parent
+		let _ = Module::<T>::add_delegation(
+			RawOrigin::Signed(parent_acc_id.clone()).into(),
+			delegation_id,
+			root_id,
+			parent,
+			delegation_acc_id.clone(),
+			permissions,
+			sig,
+		)?;
+
+		// only put in a leaf in the the first iteration
+		leaf = leaf.or(Some((
+			delegation_acc_public,
+			delegation_acc_id,
+			delegation_id,
+		)));
+	}
+	// if we didn't add children, return the parent
+	let (leaf_acc_public, leaf_acc_id, leaf_id) =
+		leaf.unwrap_or((parent_acc_public, parent_acc_id, parent_id));
```

Comment:
This should never happen and we probably should make sure it doesn't. If we have level = 10 and `children_per_level` = 0 the delegation tree will have no levels at all...

maybe `children_per_level` should be a [NonZeroInt](https://doc.rust-lang.org/std/num/struct.NonZeroU64.html)? But that's maybe to fancy...

Reply:
I know i wrote that. 😅

Reply:
I actually went with that so it's more clear. Unfortunately, the constructor wraps it into an option. See e91528a + f0679d1